### PR TITLE
Swagger (WIP)

### PR DESCRIPTION
**Pull request by ikitommi**
Initial implementation for #19. 

* Works with spec
* schema requires https://github.com/metosin/schema-tools/issues/38
* docs, later

Let's merge this now.

### CODE REVIEW

File: modules/reitit-schema/src/reitit/coercion/schema.cljc [hunk 1]
```diff
@@ -44,10 +46,27 @@
   (reify coercion/Coercion
     (-get-name [_] :schema)
     (-get-options [_] opts)
-    (-get-apidocs [_ _ {:keys [parameters responses] :as info}]
-      (cond-> (dissoc info :parameters :responses)
-              parameters (assoc ::parameters parameters)
-              responses (assoc ::responses responses)))
+    (-get-apidocs [this type {:keys [parameters responses]}]
```
Comment (miikka) [inline modules/reitit-schema/src/reitit/coercion/schema.cljc]: type shadows a clojure.core name

File: modules/reitit-spec/src/reitit/coercion/spec.cljc [hunk 2]
```diff
@@ -67,20 +68,27 @@
...
+                          (for [[k v] parameters]
+                            [k (coercion/-compile-model this v nil)]))})
+                     (if responses
+                       {::swagger/responses
+                        (into
+                          (empty responses)
+                          (for [[k response] responses
+                                :let [response (set/rename-keys response {:body :schema})]]
+                            [k (update response :schema #(coercion/-compile-model this % nil))]))})))
+        (throw
```
Comment (miikka) [inline modules/reitit-spec/src/reitit/coercion/spec.cljc]: This code looks exactly like the code for Schema, so would it make sense to de-duplicate it?

File: modules/reitit-schema/src/reitit/coercion/schema.cljc [hunk 3]
```diff
@@ -44,10 +46,27 @@
...
+                     (if responses
+                       {::swagger/responses
+                        (into
+                          (empty responses)
+                          (for [[k response] responses
+                                :let [response (set/rename-keys response {:body :schema})]]
+                            [k (update response :schema #(coercion/-compile-model this % nil))]))})))
+        (throw
+          (ex-info
+            (str "Can't produce Schem apidocs for " type)
```
Comment (miikka) [inline modules/reitit-schema/src/reitit/coercion/schema.cljc]: typo: "Schem".

File: modules/reitit-swagger/src/reitit/swagger.cljc [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+(s/def ::spec (s/keys :opt-un [::swagger ::no-doc ::tags ::summary ::description]))
+
+(def swagger-feature
+  "Feature for handling swagger-documentation for routes.
+  Works both with Middleware & Interceptors. Does not participate
+  in actual request processing, just provides specs for the extra
+  valid keys for the route data. Should be accompanied by a
+  [[swagger-spec-handler]] to expose the swagger spec.
+
+  Swagger-spesific keys:
```
Comment (miikka) [inline modules/reitit-swagger/src/reitit/swagger.cljc]: Typo: `spesific` -> `specific`

File: modules/reitit-swagger/src/reitit/swagger.cljc [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+(s/def ::description string?)
+
+(s/def ::swagger (s/keys :req-un [::id]))
+(s/def ::spec (s/keys :opt-un [::swagger ::no-doc ::tags ::summary ::description]))
+
+(def swagger-feature
+  "Feature for handling swagger-documentation for routes.
+  Works both with Middleware & Interceptors. Does not participate
+  in actual request processing, just provides specs for the extra
+  valid keys for the route data. Should be accompanied by a
```
Comment (miikka) [inline modules/reitit-swagger/src/reitit/swagger.cljc]: What are extra valid keys?

File: modules/reitit-swagger/src/reitit/swagger.cljc [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+                                    (-> parameters :query :y)})}}]]"
+  {:name ::swagger
+   :spec ::spec})
+
+(defn swagger-spec-handler
+  "Ring handler to emit swagger spec."
+  [{:keys [::r/router ::r/match :request-method]}]
+  (let [{:keys [id] :as swagger} (-> match :result request-method :data :swagger)
+        swagger (set/rename-keys swagger {:id :x-id})]
+    (if id
```
Comment (miikka) [inline modules/reitit-swagger/src/reitit/swagger.cljc]: These long single-branch ifs are really hard to parse

File: modules/reitit-swagger/src/reitit/swagger.cljc [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+      (let [paths (->> router
+                       (r/routes)
+                       (filter #(-> % second :swagger :id (= id)))
+                       (map (fn [[p _ c]]
+                              [p (some->> c
+                                          (keep
```
Comment (miikka) [inline modules/reitit-swagger/src/reitit/swagger.cljc]: You could turn this `keep` into a transducer and then do `(some->> c (into {} (keep f)))`. Furthermore, the upper-level could also be converted to a transducer. For clarity, I'd either break this function that is parameter to map (`(fn [[p _ c]] ...)`) or keep's parameter into its own function.

File: modules/reitit-schema/src/reitit/coercion/schema.cljc (hunk 1 above)
Comment (ikitommi) [reply modules/reitit-schema/src/reitit/coercion/schema.cljc]: true, it's `spesification` now.

File: modules/reitit-spec/src/reitit/coercion/spec.cljc (hunk 2 above)
Comment (ikitommi) [reply modules/reitit-spec/src/reitit/coercion/spec.cljc]: Schema is not implemented yet, might look different in the end. Marked

`;; TODO: this looks identical to spec, refactor when schema is done.`

File: modules/reitit-schema/src/reitit/coercion/schema.cljc (hunk 3 above)
Comment (ikitommi) [reply modules/reitit-schema/src/reitit/coercion/schema.cljc]: fixed

File: modules/reitit-swagger/src/reitit/swagger.cljc (hunk 4 above)
Comment (ikitommi) [reply modules/reitit-swagger/src/reitit/swagger.cljc]: fixed

File: modules/reitit-swagger/src/reitit/swagger.cljc (hunk 5 above)
Comment (ikitommi) [reply modules/reitit-swagger/src/reitit/swagger.cljc]: fixed

File: modules/reitit-swagger/src/reitit/swagger.cljc (hunk 6 above)
Comment (ikitommi) [reply modules/reitit-swagger/src/reitit/swagger.cljc]: fixed

File: modules/reitit-swagger/src/reitit/swagger.cljc (hunk 7 above)
Comment (ikitommi) [reply modules/reitit-swagger/src/reitit/swagger.cljc]: My transducer-vision is still blurry.

**Comment by ikitommi** [comment]
Thanks! Looks much better now.
